recipes/botan/all/conanfile.py

```diff
@@ -140,59 +145,108 @@ def requirements(self):
 
-        elif compiler == 'clang' and compiler.libcxx not in ['libstdc++11', 'libc++']:
+        if compiler == 'clang' and compiler.libcxx not in ['libstdc++11', 'libc++']:
             raise ConanInvalidConfiguration(
                 'Using Botan with Clang on Linux requires either "compiler.libcxx=libstdc++11" ' \
                 'or "compiler.libcxx=libc++"')
 
         # Some older compilers cannot handle the amalgamated build anymore
         # See also https://github.com/randombit/botan/issues/2328
-        if tools.Version(self.version) >= '2.14.0' and self.options.amalgamation:
-            if (compiler == 'apple-clang' and version < '10') or \
-               (compiler == 'gcc' and version < '8') or \
-               (compiler == 'clang' and version < '7'):
+        if Version(self.version) >= '2.14.0' and self.options.amalgamation:
+            if (compiler == 'apple-clang' and compiler_version < '10') or \
+               (compiler == 'gcc' and compiler_version < '8') or \
+               (compiler == 'clang' and compiler_version < '7'):
                 raise ConanInvalidConfiguration(
-                    'botan amalgamation is not supported for {}/{}'.format(compiler, version))
+                    'botan amalgamation is not supported for {}/{}'.format(compiler, compiler_version))
 
         if self.options.get_safe("single_amalgamation", False) and not self.options.amalgamation:
             raise ConanInvalidConfiguration("botan:single_amalgamation=True requires botan:amalgamation=True")
 
+    def layout(self):
+        basic_layout(self, src_folder="src")
+
     def source(self):
-        tools.get(**self.conan_data['sources'][self.version], strip_root=True, destination=self._source_subfolder)
+        get(self, **self.conan_data["sources"][self.version], strip_root=True)
+
+    @property
+    def _cxxflags(self):
+        global_cxxflags = " ".join(self.conf.get("tools.build:cxxflags", default=[], check_type=list))
+        env_cxxflags = VirtualBuildEnv(self).vars().get("CXXFLAGS")
+        if env_cxxflags is None:
+            env_cxxflags = os.getenv("CXXFLAGS", "")
+        cxxflags = f"{env_cxxflags} {global_cxxflags}".strip()
+        return cxxflags if cxxflags else None
+
+    def generate(self):
+        if is_msvc(self):
+            ms = VCVars(self)
+            ms.generate()
+
+        # This is to work around botan's configure script that *replaces* its
+        # standard (platform dependent) flags in presence of an environment
+        # variable ${CXXFLAGS}. Most notably, this would build botan with
+        # disabled compiler optimizations.
+        self._extra_cxxflags = self._cxxflags
+        self.buildenv.unset('CXXFLAGS')
+        VirtualBuildEnv(self).generate()
 
     def build(self):
-        for patch in self.conan_data.get('patches', {}).get(self.version, []):
-            tools.patch(**patch)
-        with tools.chdir(self._source_subfolder):
+        apply_conandata_patches(self)
```

Comment:
Maybe a Conan expert can answer this with a reason why:
Is `build()` or `source()` the best place to put `apply_conandata_patches()`?

The Boost recipe calls it `source()`, and it make sense to me conceptually and from what is written in https://docs.conan.io/2/reference/conanfile/methods/source.html. But in most recipe it seems to be called in the `build()` method.

Reply:
Hi! Good question. The original idea is that patching in the source method should be done for cases where the patch is common to every version, and in build when you need to differenciate per configuration ([See here for an example in the docs](https://docs.conan.io/2/examples/tools/files/patches/patch_sources.html))

But when dealing with CCI patches, you might find some relaxing of thos ideas, and people either patch in `source()` because it's the "correct" thing to do, or fall back to patching in `build()` so they don't need to think about those cases, as patching in `build()` would always be correct